pkg/standalone-utils/contracts/relayer/EulerWrapping.sol

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-interfaces/contracts/standalone-utils/IEulerToken.sol";
+import "@balancer-labs/v2-interfaces/contracts/vault/IVault.sol";
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/Address.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/math/FixedPoint.sol";
+
+import "./IBaseRelayerLibrary.sol";
+
+/**
+ * @title EulerWrapping
+ * @notice Allows users to wrap and unwrap Euler tokens
+ * @dev All functions must be payable so they can be called from a multicall involving ETH
+ */
+abstract contract EulerWrapping is IBaseRelayerLibrary {
+    using Address for address payable;
+    using SafeERC20 for IERC20;
+    using FixedPoint for uint256;
+
+    address private immutable _eulerProtocol;
+
+    /**
+     * @dev Euler protocol needs to be approved to transer mainToken
+     * @param eulerProtocol - the address of the euler protocol
+     */
+    constructor(address eulerProtocol) {
```

Comment:
By any chance, is there a way to get the address to the protocol from the tokens inside `wrapEuler`?

Looking at the other relayers, they can call `approve` just with the parameters of the `wrap` function, without requiring any pre-saved references.

The reason I'm asking this is that it'd be great if we could remove this constructor argument for maintainability reasons. E.g. if the euler protocol was upgraded and the address changed, we'd need to redeploy our relayer (I'm not sure if that's even possible, but in any case if there was a way to avoid the constructor argument it'd be desirable).

Reply:
You could pass it into the `wrapEuler` function directly, so that you don't have to store it. That would be annoying, but it's Euler's fault for not letting you get it from the tokens :)

That would get rid of the dependency and need to know it at deployment. (And be resilient if it ever changed.)

Reply:
Based on the information I have gathered I don't see a way. I think many of the other relaysers have the case, that the wrappedToken contract holds the the MainToken. For euler a seperate Contract (EULER_MAINNET) holds all funds. This address is stored in their proxy contract at slot 0. See: https://github.com/euler-xyz/euler-contracts/blob/master/contracts/Proxy.sol#L6

Reply:
I agree with @EndymionJkb: we should add the Euler address as an argument to the `wrapEuler()` and `unwrapEuler()` functions. That would even be sufficient for dealing with any potential forks of Euler down the road, which would have different protocol addresses.

Reply:
It's unfortunate that Euler makes you do that (though I'm sure they have a good reason for it); it was just really messy to have to change files of unrelated protocols (even just mocks), and introduce this explicit dependency

Reply:
Addressed in https://github.com/balancer/balancer-v2-monorepo/pull/2327/commits/ea0341b95381f1f0006e96015e30970f4c82fa6c.